File: piper/extensions/chrome/src/engine/statements/is_html_element_exist.js
import { stringify } from "../util"


export default async function(statement, onError) {
    try {
        let els = await this.executeStatement(stringify(statement.pointer))
        if (!(els instanceof Array)) {
            els = [els]
        }

        let tab = await this.browser.getCurrentTab()
        let out = false
        for (let el of els) {
            let xpath = el.attrs.xpath || "null"
            let isExist = await chrome.scripting.executeScript({
                target: { tabId: tab.id },
                func: (xpath, el) => {
                    let foundEl = document.querySelectorAll(el.selector)[el.elIndex]
                    if (!foundEl) {
                        el = document.evaluate(xpath, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue
                    }
                    return (el !== undefined && el !== null)
                },
                args: [xpath, el],
                world: "MAIN"
            })
            out = isExist[0].result
            if (!out) break
        }
        return out
    } catch (e) {
        onError(e)
        return false
    }
}

File: piper/android/app/src/main/java/io/gopiper/piper/engine/statements/is_html_element_exist.kt
package io.gopiper.piper.engine.statements

import kotlinx.coroutines.delay
import io.gopiper.piper.cheese.js
import io.gopiper.piper.engine.Piper
import io.gopiper.piper.engine.getXpath
import io.gopiper.piper.engine.model.HTMLElement
import me.gumify.hiper.util.onUiThread
import org.json.JSONObject
import java.lang.Exception

suspend fun Piper.executeIsHtmlElementExist(statement: JSONObject, onError: (e: Exception) -> Unit): Boolean {
    try {
        delay(1000)
        val elsObj = executeStatement(statement["pointer"].toString())

        val elements = if (elsObj is List<*>) {
            elsObj as List<HTMLElement>
        } else {
            listOf(elsObj as HTMLElement)
        }

        var out = false
        for (el in elements) {
            var isExist: Boolean? = null
            val xpath = el.attrs.getXpath()
            onUiThread {
                browser.currentTab.value?.webView?.js("""
                var el = document.querySelectorAll("${el.selector}")[${el.elIndex}];
                if (el === undefined) {
                    el = document.evaluate('${xpath}', document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
                }
                return (el !== undefined && el !== null);
            """.trimIndent()) {
                    isExist = it == "true"
                }
            }
            while (isExist == null) {
                if (isTerminated) break
                delay(5)
            }
            out = isExist ?: false
            if (isExist == false) {
                break
            }
        }
        return out
    } catch (e: Exception) {
        onError(e)
        return false
    }
}